storage.go

```diff
@@ -2,35 +2,154 @@ package bos
+		if !s.features.VirtualDir {
+			return
+		}
+
+		rp += "/"
+		o = s.newObject(true)
+		o.Mode |= ModeDir
+	} else {
+		o = s.newObject(true)
+		o.Mode |= ModeRead
+	}
+	o.ID = rp
+	o.Path = path
+
+	return o
 }
 
 func (s *Storage) delete(ctx context.Context, path string, opt pairStorageDelete) (err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+
+	if opt.HasObjectMode && opt.ObjectMode.IsDir() {
+		if !s.features.VirtualDir {
+			err = services.PairUnsupportedError{Pair: ps.WithObjectMode(opt.ObjectMode)}
+			return
+		}
+		rp += "/"
+	}
+
+	err = s.client.DeleteObject(s.bucket, rp)
+	if err != nil {
+		return err
+	}
+
+	return nil
 }
 
 func (s *Storage) list(ctx context.Context, path string, opt pairStorageList) (oi *ObjectIterator, err error) {
 	panic("not implemented")
 }
 
 func (s *Storage) metadata(opt pairStorageMetadata) (meta *StorageMeta) {
-	panic("not implemented")
+	meta = NewStorageMeta()
+	meta.Name = s.bucket
+	meta.WorkDir = s.workDir
+	return meta
 }
 
 func (s *Storage) read(ctx context.Context, path string, w io.Writer, opt pairStorageRead) (n int64, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+
+	output, err := s.client.BasicGetObject(s.bucket, rp)
+	if err != nil {
+		return 0, err
+	}
+
+	rc := output.Body
+	if opt.HasIoCallback {
+		rc = iowrap.CallbackReadCloser(rc, opt.IoCallback)
+	}
+
+	return io.Copy(w, rc)
 }
 
 func (s *Storage) stat(ctx context.Context, path string, opt pairStorageStat) (o *Object, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+
+	if opt.HasObjectMode && opt.ObjectMode.IsDir() {
+		if !s.features.VirtualDir {
+			err = services.PairUnsupportedError{Pair: ps.WithObjectMode(opt.ObjectMode)}
+			return nil, err
+		}
+
+		rp += "/"
+	}
+
+	output, err := s.client.GetObject(s.bucket, rp, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	o = s.newObject(true)
+	o.ID = rp
+	o.Path = path
+
+	if opt.HasObjectMode && opt.ObjectMode.IsDir() {
+		o.Mode |= ModeDir
+	} else {
+		o.Mode |= ModeRead
+	}
+
+	o.SetContentLength(output.ContentLength)
+	lastModified, err := time.Parse("Mon, 02 Jan 2006 15:04:05 GMT", output.LastModified)
+	if err != nil {
+		return nil, err
+	}
+	o.SetLastModified(lastModified)
+
+	if output.ContentType != "" {
+		o.SetContentType(output.ContentType)
+	}
+	if output.ETag != "" {
+		o.SetEtag(output.ETag)
+	}
+
+	var sm ObjectSystemMetadata
+	if v := output.StorageClass; v != "" {
+		sm.StorageClass = v
+	}
+
+	o.SetSystemMetadata(sm)
+
+	return
 }
 
 func (s *Storage) write(ctx context.Context, path string, r io.Reader, size int64, opt pairStorageWrite) (n int64, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+
+	r = io.LimitReader(r, size)
+	if opt.HasIoCallback {
+		r = iowrap.CallbackReader(r, opt.IoCallback)
+	}
+
+	bodyStream, err := bce.NewBodyFromFile(rp)
+	if err != nil {
+		return 0, err
+	}
+	args := &api.PutObjectArgs{}
+
+	_, err = s.client.PutObject(s.bucket, rp, bodyStream, args)
```

Comment:
For `PutObject`, the third parameter `body` is the object content body, but `rp` is the target object instead of source data file.

Reply:
Maybe we can use `NewBodyFromSizedReader(r, size)` here and check if `opt.ContentMD5`(if has) is equal to `body.ContentMD5()`?